tree/treeplayer/inc/ROOT/TDFActionHelpers.hxx

```diff
@@ -92,6 +93,30 @@ public:
    ULong64_t &PartialUpdate(unsigned int slot);
 };
 
+template<typename ProxiedVal_t>
+class ReportHelper {
+   const std::shared_ptr<TCutFlowReport> fReport;
+   std::weak_ptr<ProxiedVal_t> fProxiedWPtr;
+   bool fReturnEmptyReport;
+public:
+   using BranchTypes_t = TypeList<>;
+   ReportHelper(const std::shared_ptr<TCutFlowReport> &report, const std::shared_ptr<ProxiedVal_t> &pp, bool emptyRep)
+   :fReport(report), fProxiedWPtr(pp), fReturnEmptyReport(emptyRep) {};
+   ReportHelper(ReportHelper &&) = default;
+   ReportHelper(const ReportHelper &) = delete;
+   void InitSlot(TTreeReader *, unsigned int) {}
+   void Exec(unsigned int /* slot */ ) {}
+   void Initialize() { /* noop */ }
+   void Finalize()
+   {
+      // We need the weak_ptr in order to avoid crashes at tear down
```

Comment:
`TLoopManager` keeps the whole graph alive, and without `TLoopManager` the event loop cannot run.
What is the scenario in which the weak_ptr could be expired?

Reply:
if the whole chain goes out of scope, right?

Reply:
The chain only goes out of scope if the head node goes out of scope, in which case all hell breaks loose anyway